src/main/java/juja/microservices/gamification/service/AchievementService.java

```diff
@@ -220,97 +219,92 @@ private void addThanksKeeperScheduled() {
         List<String> achievementIds;
         if (achievements.isEmpty()) {
             achievementIds = createThanksKeeperAchievements();
+            logger.info("Added new THANKS KEEPER achievement. Ids: {}", achievementIds);
         } else {
-            achievementIds = getIdsCreatedAchievement(achievements);
+            achievementIds = achievements.stream()
+                    .map(Achievement::getId)
+                    .collect(Collectors.toList());
+            logger.info("Returned already created THANKS KEEPER achievements in current week {}", achievementIds);
         }
         return achievementIds;
     }
 
     private List<Achievement> getThanksKeeperAchievements() {
-        logger.debug("Send request to repository: get all thanks_keepers achievements for current week");
+        logger.debug("Send request to repository: get already created THANKS KEEPER achievements for current week");
         List<Achievement> achievements = achievementRepository.getAllThanksKeepersAchievementsCurrentWeek();
-        logger.debug("Received list of achievements, size = {}", achievements.size());
+        logger.debug("Received list of THANKS KEEPER achievements, size = {}", achievements.size());
         return achievements;
     }
 
     private List<String> createThanksKeeperAchievements() {
-        List<String> result = new ArrayList<>();
         logger.debug("Request to keepers repository: get all active keepers");
         List<KeeperDTO> keepers = keeperService.getKeepers();
         logger.debug("Received list of active keepers, size = {} ", keepers.size());
+        List<String> result = new ArrayList<>();
         if (keepers.isEmpty()) {
             logger.debug("No any active keepers received from user service");
         } else {
-            logger.debug("Sending 'Thanks Keeper' achievements to repository");
-            for (KeeperDTO keeper : keepers) {
-                List<String> directions = keeper.getDirections();
-                directions.forEach(direction -> {
-                    String description = String.format(THANKS_DESCRIPTION, direction);
-                    Achievement achievement = new Achievement(
-                            systemFrom, keeper.getUuid(), KEEPER_THANKS, description, AchievementType.THANKS_KEEPER);
-                    result.add(achievementRepository.addAchievement(achievement));
-                });
-            }
-            logger.debug("Received ids from repository: {}", result);
+            result = keepers.stream()
+                    .map(this::prepareThanksKeeperAchievements)
+                    .flatMap(Collection::stream)
```

Comment:
Why did you use flatMap ?
Please describe this case for me :)

Reply:
FlatMap converts each element of the stream into a stream of other objects.
There is we have a stream of KeeperDTO objects, each has  nested List of directions, so we prepare list of achivement based on uuid of keeper and direction. flatMap produce new stream from prepared list of achivements and in the next step we work with stream of Achivement objects.

Reply:
Thanks. I got it.